activation/nipost.go

```diff
@@ -346,50 +337,74 @@ func (nb *NIPostBuilder) submitPoetChallenge(
 	signature types.EdSignature,
+	certifier certifierService,
 ) error {
 	logger := nb.log.With(
 		log.ZContext(ctx),
 		zap.String("poet", client.Address()),
 		log.ZShortStringer("smesherID", nodeID),
 	)
 
-	logger.Debug("querying for poet pow parameters")
-	powCtx, cancel := withConditionalTimeout(ctx, nb.poetCfg.RequestTimeout)
-	defer cancel()
-	powParams, err := client.PowParams(powCtx)
-	if err != nil {
-		return &PoetSvcUnstableError{msg: "failed to get PoW params", source: err}
+	// FIXME: remove support for deprecated poet PoW
+	auth := PoetAuth{
+		PoetCert: certifier.GetCertificate(client.Address()),
 	}
-
-	logger.Debug("doing pow with params", zap.Any("pow_params", powParams))
-	startTime := time.Now()
-	nonce, err := shared.FindSubmitPowNonce(
-		ctx,
-		powParams.Challenge,
-		challenge,
-		nodeID.Bytes(),
-		powParams.Difficulty,
-	)
-	metrics.PoetPowDuration.Set(float64(time.Since(startTime).Nanoseconds()))
-	if err != nil {
-		return fmt.Errorf("running poet PoW: %w", err)
+	if auth.PoetCert == nil {
+		logger.Info("missing poet cert - falling back to PoW")
+		powCtx, cancel := withConditionalTimeout(ctx, nb.poetCfg.RequestTimeout)
+		defer cancel()
+		powParams, err := client.PowParams(powCtx)
+		if err != nil {
+			return &PoetSvcUnstableError{msg: "failed to get PoW params", source: err}
+		}
+		logger.Debug("doing pow with params", zap.Any("pow_params", powParams))
+		startTime := time.Now()
+		nonce, err := shared.FindSubmitPowNonce(
+			ctx,
+			powParams.Challenge,
+			challenge,
+			nodeID.Bytes(),
+			powParams.Difficulty,
+		)
+		metrics.PoetPowDuration.Set(float64(time.Since(startTime).Nanoseconds()))
+		if err != nil {
+			return fmt.Errorf("running poet PoW: %w", err)
+		}
+		auth.PoetPoW = &PoetPoW{
+			Nonce:  nonce,
+			Params: *powParams,
+		}
+	} else {
+		logger.Info("registering with a certificate", zap.Binary("cert", auth.PoetCert.Data))
 	}
 
 	logger.Debug("submitting challenge to poet proving service")
 
 	submitCtx, cancel := withConditionalTimeout(ctx, nb.poetCfg.RequestTimeout)
 	defer cancel()
-	round, err := client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID, PoetPoW{
-		Nonce:  nonce,
-		Params: *powParams,
-	})
-	if err != nil {
-		return &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
-	}
+	var (
+		round *types.PoetRound
+		err   error
+	)
 
+	for try := 0; try < 2; try++ {
+		round, err = client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID, auth)
+		if err == nil {
+			break
+		}
+		if errors.Is(err, ErrUnathorized) && try == 0 {
+			logger.Warn("failed to submit challenge as unathorized - recertifying", zap.Error(err))
+			auth.PoetCert, err = certifier.Recertify(ctx, client)
+			if err != nil {
+				return &PoetSvcUnstableError{msg: "failed to regenerate poet certificate", source: err}
+			}
+		} else {
+			return &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
+		}
+	}
```

Comment:
I think we had this discussion before; the loop doesn't immediately make sense when reading the code. How about

```suggestion
	var (
		round *types.PoetRound
		err   error
	)
	round, err = client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID, auth)
	switch {
	case errors.Is(err, ErrUnathorized):
		logger.Warn("failed to submit challenge as unathorized - recertifying", zap.Error(err))
		auth.PoetCert, err = certifier.Recertify(ctx, client)
		if err != nil {
			return &PoetSvcUnstableError{msg: "failed to regenerate poet certificate", source: err}
		}
		round, err = client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID, auth)
		if err != nil {
			return &PoetSvcUnstableError{msg: "failed to regenerate poet certificate", source: err}
		}
	case err != nil:
		return &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
	default: // err == nil
	}
```

or even better, remove the logic for getting the certificate and adding it to `auth` from the builder and let the `client.Submit` method handle it by itself?